imodels/tree/hierarchical_shrinkage.py

```diff
@@ -225,14 +226,36 @@ def __init__(self, estimator_: BaseEstimator = None,
         #     raise Warning('Passed an already fitted estimator,'
         #                   'but shrinking not applied until fit method is called.')
 
-    def fit(self, X, y, *args, **kwargs):
+    def fit_slow(self, X, y, *args, **kwargs):
+        s = time.time()
         self.scores_ = []
         for reg_param in self.reg_param_list:
             est = HSTreeClassifier(deepcopy(self.estimator_), reg_param)
             cv_scores = cross_val_score(est, X, y, cv=self.cv, scoring=self.scoring)
             self.scores_.append(np.mean(cv_scores))
         self.reg_param = self.reg_param_list[np.argmax(self.scores_)]
         super().fit(X=X, y=y, *args, **kwargs)
+        e = time.time() - s
+        print(f"slow time: {np.round(e, 2)}")
+
+    def fit(self, X, y, *args, **kwargs):
+        s = time.time()
+        self.scores_ = [[] for _ in self.reg_param_list]
+        kf = KFold(n_splits=self.cv)
```

Comment:
I'd recommend a neat trick I've implemented in AutoGluon when the base estimator is a forest: No need to use KFold, just get the out-of-bag prediction probabilities and score them!

https://github.com/awslabs/autogluon/blob/master/tabular/src/autogluon/tabular/models/rf/rf_model.py#L262-L334

This allows you to
1. Avoid having to do KFold, only need to fit 1 forest
2. No need to fit again on all the data once the optimal value is found, since you already used all the data when fitting the model to tune the reg_param.

I understand if this is complicated logic for this repo, so I can implement support for it within AutoGluon specifically if it isn't included in this repo.

(ditto for regression)

Reply:
This is a really nice trick! I agree this is a little complicated to add here for now (esp. since we focus mostly on the single-tree), but would be a good add eventually.